art/attacks/evasion/steal_now_attack_latter.py

```diff
@@ -0,0 +1,747 @@
+        self.k = k
+        self.patch_list = [TileObj(tile_size=tile_size, device=device)] * self.k
+
+    def insert(self, target: TileObj) -> None:
+
+        if target.bcount < self.threshold:
+            return
+
+        prev = self.patch_list
+        out = []
+        for k_it in range(self.k):
+            if target.compare(prev[k_it]):
+                out.append(target)
+                out = out + prev[k_it:]
+                break
+            else:
+                out.append(prev[k_it])
+
+        self.patch_list = out[:self.k]
+
+    def pop(self) -> None:
+        out = self.patch_list[1:] + [TileObj(tile_size=self.strides, device=self.device)]
+        self.patch_list = out
+
+
+class SNAL(EvasionAttack):
+    """
+    Steal Now and Attack Later
+    | Paper link: https://arxiv.org/abs/2404.15881
+    """
+
+    attack_params = EvasionAttack.attack_params + [
+        "eps",
+        "max_iter",
+        "num_grid",
+        "batch_size",
+    ]
+
+    _estimator_requirements = ()
+
+    def __init__(
+        self,
+        estimator: "torch.nn.Module",
+        eps: float,
+        max_iter: int,
+        num_grid: int,
+    ) -> None:
+        """
+        Create a SNAL attack instance.
+        :param estimator: A trained YOLOv8 model or other models with the same output format
+        :param eps: Maximum perturbation that the attacker can introduce.
+        :param max_iter: The maximum number of iterations.
+        :param num_grid: The number of grids for width and high dimension.
+        """
+        super().__init__(estimator=estimator)
+        self.eps = eps
+        self.max_iter = max_iter
+        self.num_grid = num_grid
+        self.batch_size = 1
+        self.candidates = None
+        self.threshold_objs = 1 # the expect number of objects
+        self._check_params()
+
+    def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> np.ndarray:
+        """
+        Generate adversarial samples and return them in an array.
+        :param x: An array with the original inputs to be attacked.
+        :param y: Not used.
+        :return: An array holding the adversarial examples.
+        """
+
+        # Compute adversarial examples with implicit batching
+        x_adv = x.copy()
+        for batch_id in range(int(np.ceil(x_adv.shape[0] / float(self.batch_size)))):
+            batch_index_1 =  batch_id * self.batch_size
+            batch_index_2 = min((batch_id + 1) * self.batch_size, x_adv.shape[0])
+            x_batch = x_adv[batch_index_1:batch_index_2]
+            x_adv[batch_index_1:batch_index_2] = self._generate_batch(x_batch)
+
+        return x_adv
+
+    def _generate_batch(self, x_batch: np.ndarray, y_batch: Optional[np.ndarray] = None) -> np.ndarray:
+        """
+        Run the attack on a batch of images.
+        :param x_batch: A batch of original examples.
+        :param y_batch: Not Used.
+        :return: A batch of adversarial examples.
+        """
```

Comment:
Add empty line for docs syntax.
```suggestion
        """
        Run the attack on a batch of images.

        :param x_batch: A batch of original examples.
        :param y_batch: Not Used.
        :return: A batch of adversarial examples.
        """
```